### DISCUSSION TITLE
`.extend` on a `ZodUnion` of objects

### DISCUSSION THREAD

**Discussion by rijenkii** [colinhacks/zod#2279] [Q&A] [answered]
Recently I really needed a method for `.extend`ing all elements of a union and returning a new union.
Because I am a newb with TypeScript, this has taken me multiple days, but I was able to concoct something that I think works:

```typescript
import * as z from "zod";

function zodUnionExtend<
  Us extends [z.AnyZodObject, z.AnyZodObject, ...z.AnyZodObject[]],
  S extends z.ZodRawShape
>(
  union: z.ZodUnion<Us>,
  shape: S
): z.ZodUnion<{
  [K in keyof Us]: z.ZodObject<
    z.objectUtil.extendShape<Us[K]["shape"], S>,
    Us[K]["_def"]["unknownKeys"],
    Us[K]["_def"]["catchall"]
  >;
}> {
  /* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
  // @ts-ignore
  return z.union(union.options.map((o) => o.extend(shape)));
}
```

[TypeScript Playground](https://www.typescriptlang.org/play#code/JYWwDg9gTgLgBAKjgQwM5wF5wGZQiOAIgwgBNCBuAKCuwFcA7AYxmAgczIFUG2GBRAB4wApg1IAeKnDhd0I4WNLoA2hgB0AQQYBPAFpkA8gCMAViJYAaTFt0HSJ81bjrXG7fqNmLMFQF0-S2k4AGU4BVFxdA17ACVkAHcQgAtkMBEqAD4ACmDGPgAuG3sePgk5TKCZVFT0opCqAEoimO5edgkAb2CVAGk4YA4AaxEdCGxZVD8W9XtHHykZGQ0IbxYuVgAbdQilFLSRctQ+vxVCGoPCQNDK4Jk5E7OAfVIRbCuzxiGGCASGXtGqCuVSWD16p0ILzeH0ITGQMCYqU2myuwUy1AAvpk4N0ZAB6JAiVCbQYwAC0pGAqGQxk2IjJDAiZJJjLgAAEYDp0qgmFBgGByUSWTA8cZkAwyTBUGSmPgQGJ4Ag8cE8Xj2VKycAAOY-KAZGR6mB0KAcDT5djZc0MdQQAV8VDqEBpbLZCCNOAAXmxEB2inE2Qu6UawcxNFlDFQ8GQnpsVuyKmCKzWMGynRQMxJoigyE22UIyEI7oxjRBSacKbTyAzwCzObzxkLcGLpZtydT6ZsmZE2dzsMbzaofka1Co4cjcGMMZIpFK7CEkVI2WQ1jTxmrtd7DaLw7D7HHTwAjDHjDa7XuVAAGPzqQMiaiqpZwAB6AH5R3v4E8AEzH0+sc8Htet73mqSyvu+EafgAzL+tr-hGKhfkBtR3lQD5gS+QA)

I tried getting return type of `.extend` through `ReturnType`, but in the end I did not succeed and just copied the signature from the source code.

**Comment by mmvsk** [comment, marked as answer]
Use [ZodIntersection](https://github.com/colinhacks/zod#intersections):

```typescript
const CircleSchema = Z.object({
    radius: Z.number(),
});

const RectangleSchema = Z.object({
    sideA: Z.number(),
    sideB: Z.number(),
});

const ShapeSchema = Z.discriminatedUnion("type", [
    CircleSchema.extend({ type: Z.literal("circle") }),
    RectangleSchema.extend({ type: Z.literal("rectangle") }),
]);

const PositionSchema = Z.object({ x: Z.number(), y: Z.number() });

const PlanObjectSchema = Z.intersection(ShapeSchema, PositionSchema);

type Shape = Z.output<typeof ShapeSchema>;
type Position = Z.output<typeof PositionSchema>;
type PlanObject = Z.output<typeof PlanObjectSchema>;

function CreatePlanObject(shape: Shape, position: Position): PlanObject {
    return {
        ...shape,
        ...position,
    };
}
```

**Reply by andho** [reply]
Sadly this will lose function of ZodObject like `pick` and `omit` according to docs, but seems like the only way for now.
